### ISSUE TITLE
Clobbering of slots in MT protocol handling

### ISSUE THREAD

**Issue by mapinguari** [koreader/koreader-base#1437]
Hi all,

If a device has multiple input fds each of which is conforming to the MT B protocol you end up with clobbering of MT slots.

For example, if I have two touch devices as above.

fd 1 produces:
`ABS_MT_SLOT 0
ABS_MT_TRACKING_ID 45
ABS_MT_POSITION_X 0
ABS_MT_POSITION_Y 0
SYN_REPORT
`
then some time later produces

`ABS_MT_POSITION_X 100
ABS_MT_POSITION_Y 100
SYN_REPORT`

and fd2 produces:
`ABS_MT_SLOT 0
ABS_MT_TRACKING_ID 6
ABS_MT_POSITION_X 1000
ABS_MT_POSITION_Y 1000
SYN_REPORT`

From two distinct touch events, say, one from a finger and one from a pen tool then depending on the order of the reads from the fds you can end up with any permutation of touches at (0,0), (100,100) and (1000,1000).

The input functions need to make sure they are keeping track of which fd was the last to be read from, what the last slot it was producing events for was and if it is a different fd to ensure an ABS_MT_SLOT event for the current fd's last slot is inserted into the table of events being generated. It probably also needs to maintain a mapping between an individual fd's slot number and a "globally unique" slot number so you don't end up with clobbering of slots between different input devices.

I don't think this is a driver issue, each individual driver is doing the correct thing and respecting the MT protocol, it is just when you combine multiple input sources together into a combined stream you end up with unfortunate interlacing of the event streams.

I am not sure how much appetite you have to look at this tho. 

Any feedback apprecited!

**Comment by NiLuJe** [comment]
It's a simplification, as this should be an exceedingly rare occurrence in practice.

(Also, we also clobber slots on some Kobo protocols later, anyway).

**Comment by mapinguari** [comment]
Fair enough.

I will do by best to handle it further downstream. 

Want me to close this?

**Comment by NiLuJe** [comment]
> Want me to close this?

Not necessarily, base's issue tracker isn't crowded, so leaving this open isn't really problematic.

As far as I'm concerned, this would be fairly low on the priority list, even if I actually had time to work on this (spoiler: I don't have *any* time ;p).

**Comment by NiLuJe** [comment]
Could *possibly* be as simple as passing along the fd number in Lua-land and simply using  `fd + ABS_MT_SLOT` in GestureManager, as we don't actually care about the slots value (in fact, some Kobo devices don't even emit a slot 0).

**Comment by mapinguari** [comment]
Yeah I did think about something along those lines, but I am not entirely sure what goes on in KOReader wrt slots and ids once they go into gesture detector.

I am just getting some funny double free exceptions from somewhere and it seems to be related to touch and tool events mingling.

**Comment by NiLuJe** [comment]
Might need to be slightly more complex, because of the initial state in https://github.com/koreader/koreader/blob/19a607b5484299623cc57563f7a4baa3f49faff5/frontend/device/input.lua#L220-L226

If we had a guarantee that the first event device we open is the touch screen, we might be able to use the [`inputfds`](https://github.com/koreader/koreader-base/blob/1432b9298e8a2ab00aaa2cc52ca14fa90665c70f/input/input.c#L50) index instead, but we don't.

**Comment by NiLuJe** [comment]
It would *probably* not be too gnarly to deal with, I just have zero time to devote to this (or anything else, really :/) for the foreseeable future.

**Comment by mapinguari** [comment]
ok cool.

I will leave it be.

**Comment by NiLuJe** [comment]
There might be a similar issue with cross-devices `ABS_MT_TRACKING_ID` ids, too.

**Comment by mapinguari** [comment]
From my little bit of watching the event streams on this one device then yes, it looks rarer. One device always uses the same `ABS_MT_TRACKING_ID` 0, and the other just continuously increments it.

**Comment by NiLuJe** [comment]
As far as passing along the fd info, it should only take some minor contortions to stuff it in the [table](https://github.com/koreader/koreader-base/blob/1432b9298e8a2ab00aaa2cc52ca14fa90665c70f/input/input.c#L274-L300) ;).

**Comment by mapinguari** [comment]
that is probably the simplest unobtrusive fix
